File: agent/collectors/hardware.py
import platform
import os
import wmi
import pythoncom
import math

def get_system_info():
    """Detailed System Info"""
    info = {
        "hostname": platform.node(),
        "os": platform.system(),
        "release": platform.release(),
        "version": platform.version(),
        "architecture": platform.machine(),
        "windows_build": "",
        "caption": "",
        "bios_serial": "",
        "last_user": os.getlogin() if hasattr(os, 'getlogin') else ""
    }
    
    if platform.system() == "Windows":
        try:
            pythoncom.CoInitialize()
            c = wmi.WMI()
            # BIOS Serial
            for bios in c.Win32_BIOS():
                info["bios_serial"] = bios.SerialNumber
            # Windows Build and OS Caption
            for os_info in c.Win32_OperatingSystem():
                info["windows_build"] = os_info.BuildNumber
                info["caption"] = getattr(os_info, "Caption", "")
        except:
            pass
            
    return info

def collect_monitors():
    """Scan connected monitors using WMI (Win32_DesktopMonitor + WmiMonitorID)"""
    pythoncom.CoInitialize()
    monitors = []
    if platform.system() != "Windows":
        return monitors

    try:
        c_wmi = wmi.WMI(namespace="root\\wmi")
        mon_ids = c_wmi.WmiMonitorID()
        params_list = c_wmi.WmiMonitorBasicDisplayParams()

        # ดึง Resolution รวมจาก Win32_VideoController
        resolution = ""
        try:
            c_win32 = wmi.WMI()
            for vc in c_win32.Win32_VideoController():
                h = getattr(vc, "CurrentHorizontalResolution", 0)
                v = getattr(vc, "CurrentVerticalResolution", 0)
                if h and v:
                    resolution = f"{h}x{v}"
                    break
        except Exception:
            pass

        def decode_bytes(arr):
            if not arr:
                return ""
            return "".join(chr(b) for b in arr if b != 0).strip()

        for idx, mon in enumerate(mon_ids):
            try:
                serial = decode_bytes(getattr(mon, "SerialNumberID", []))
                model = decode_bytes(getattr(mon, "UserFriendlyName", []))
                manufacturer = decode_bytes(
                    getattr(mon, "ManufacturerName", [])
                )

                size_inch = 0
                size_cm = 0

                # จับคู่ params ตาม Index เดียวกันกับ mon_ids
                if idx < len(params_list):
                    p = params_list[idx]
                    w_cm = getattr(p, "MaxHorizontalImageSize", 0)  # cm
                    h_cm = getattr(p, "MaxVerticalImageSize", 0)  # cm

                    if w_cm > 0 and h_cm > 0:
                        diag_cm = math.sqrt(w_cm**2 + h_cm**2)
                        size_cm = round(diag_cm)
                        size_inch = round(
                            diag_cm / 2.54
                        )  # แปลงเซนติเมตรเป็นนิ้ว

                if model:  # เพิ่มเฉพาะรายการที่มีชื่อรุ่น
                    monitors.append(
                        {
                            "serial": serial or "N/A",
                            "model": model,
                            "manufacturer": manufacturer,
                            "size_inch": size_inch,  # นิ้ว (เช่น 24, 27)
                            "size_cm": size_cm,  # เซนติเมตร
                            "resolution": resolution,
                        }
                    )
            except Exception as e:
                print(f"[Monitor] Error reading monitor entry: {e}")

    except Exception as e:
        print(f"[Monitor] WMI namespace error: {e}")
        # Fallback: ใช้ Win32_DesktopMonitor
        try:
            c = wmi.WMI()
            for mon in c.Win32_DesktopMonitor():
                name = getattr(mon, "Name", "") or ""
                if name and name.strip():
                    monitors.append(
                        {
                            "serial": "N/A",
                            "model": name.strip(),
                            "manufacturer": "",
                            "size_inch": 0,
                            "size_cm": 0,
                            "resolution": "",
                        }
                    )
        except Exception as e2:
            print(f"[Monitor] Fallback error: {e2}")

    return monitors

def get_hardware_info():
    """Deep Hardware Scan using WMI"""
    pythoncom.CoInitialize()
    hw = {
        "cpu": {},
        "ram": [],
        "storage": [],
        "gpu": [],
        "motherboard": {},
        "bios": {},
        "printers": []
    }
    
    if platform.system() == "Windows":
        try:
            c = wmi.WMI()
            
            # CPU Detail
            for processor in c.Win32_Processor():
                hw["cpu"] = {
                    "name": processor.Name.strip(),
                    "cores": processor.NumberOfCores,
                    "threads": processor.NumberOfLogicalProcessors,
                    "max_speed": processor.MaxClockSpeed,
                    "l3_cache": getattr(processor, "L3CacheSize", 0)
                }
            
            # RAM Slots Detail
            for mem in c.Win32_PhysicalMemory():
                hw["ram"].append({
                    "capacity_gb": round(int(mem.Capacity) / (1024**3), 2),
                    "speed": mem.Speed,
                    "manufacturer": mem.Manufacturer.strip(),
                    "part_number": mem.PartNumber.strip(),
                    "serial": mem.SerialNumber.strip()
                })
            
            # Physical Disks with S.M.A.R.T. Predict Failure
            # Step 1: Build a mapping of InstanceName -> predict_failure from MSStorageDriver_FailurePredictStatus
            smart_failure_map = {}
            try:
                c_wmi = wmi.WMI(namespace="root\\wmi")
                for smart in c_wmi.MSStorageDriver_FailurePredictStatus():
                    inst = getattr(smart, "InstanceName", "") or ""
                    # InstanceName format: "SCSI\\DISK&...\\X&Y&Z{N}" — split by __ to get device ID prefix
                    predict = getattr(smart, "PredictFailure", False) or False
                    # Use the first segment before "__" as key for partial matching
                    key = inst.split("\\")[-1].upper() if inst else ""
                    smart_failure_map[key] = bool(predict)
            except Exception:
                pass  # WMI SMART query not supported on this machine

            for disk in c.Win32_DiskDrive():
                # Map this disk's PNPDeviceID or DeviceID against SMART map
                pnp = (disk.PNPDeviceID or "").upper()
                device_id_short = pnp.split("\\")[-1].upper() if pnp else ""

                # Try to find a matching entry in the SMART failure map
                predict_failure = False
                for key, val in smart_failure_map.items():
                    if key and (key in pnp or device_id_short in key):
                        predict_failure = val
                        break
                else:
                    # Fallback: check Win32_DiskDrive.Status field
                    status = (getattr(disk, "Status", "") or "").upper()
                    if "PRED FAIL" in status:
                        predict_failure = True

                hw["storage"].append({
                    "model": disk.Model,
                    "size_gb": round(int(disk.Size) / (1024**3), 2) if disk.Size else 0,
                    "serial": disk.SerialNumber.strip() if disk.SerialNumber else "N/A",
                    "interface": disk.InterfaceType,
                    "media_type": disk.MediaType,
                    "predict_failure": predict_failure
                })
            
            # GPU Detail
            for gpu in c.Win32_VideoController():
                try:
                    raw_ram = getattr(gpu, "AdapterRAM", 0)
                    ram_mb = round(int(raw_ram) / (1024**2), 2) if raw_ram else 0
                except (ValueError, TypeError):
                    ram_mb = 0
                hw["gpu"].append({
                    "name": getattr(gpu, "Name", "Unknown"),
                    "driver_version": getattr(gpu, "DriverVersion", "N/A"),
                    "ram_mb": ram_mb
                })
            
            # Motherboard
            for board in c.Win32_BaseBoard():
                hw["motherboard"] = {
                    "manufacturer": board.Manufacturer,
                    "product": board.Product,
                    "version": board.Version
                }

            # BIOS
            for b in c.Win32_BIOS():
                hw["bios"] = {
                    "version": b.SMBIOSBIOSVersion,
                    "serial": b.SerialNumber,
                }
            
            # Import and populate printers information
            try:
                from collectors.printers import get_printers_info
                hw["printers"] = get_printers_info()
            except Exception as pe:
                print(f"[HW] Printer Scan Error: {pe}")

        except Exception as e:
            print(f"[HW] WMI Error: {e}")
            
    return hw

File: go_agent/collectors/hardware.go
package collectors

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	"github.com/go-ole/go-ole"
	"github.com/go-ole/go-ole/oleutil"
)

// ---- WMI helpers ----

func wmiConnect() (*ole.IDispatch, *ole.IDispatch, error) {
	if err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED); err != nil {
		// already initialized in this goroutine – ok
	}
	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
	if err != nil {
		return nil, nil, err
	}
	wmi, err := unknown.QueryInterface(ole.IID_IDispatch)
	if err != nil {
		unknown.Release()
		return nil, nil, err
	}
	unknown.Release()
	serviceRaw, err := oleutil.CallMethod(wmi, "ConnectServer")
	if err != nil {
		wmi.Release()
		return nil, nil, err
	}
	return wmi, serviceRaw.ToIDispatch(), nil
}

func wmiQueryItems(service *ole.IDispatch, query string) ([]*ole.IDispatch, func()) {
	res, err := oleutil.CallMethod(service, "ExecQuery", query)
	if err != nil {
		return nil, func() {}
	}
	r := res.ToIDispatch()
	countVar, _ := oleutil.GetProperty(r, "Count")
	count := int(countVar.Val)
	items := make([]*ole.IDispatch, 0, count)
	for i := 0; i < count; i++ {
		itemRaw, err := oleutil.CallMethod(r, "ItemIndex", i)
		if err != nil {
			continue
		}
		items = append(items, itemRaw.ToIDispatch())
	}
	cleanup := func() {
		for _, it := range items {
			it.Release()
		}
		r.Release()
	}
	return items, cleanup
}

func strProp(it *ole.IDispatch, name string) string {
	v, err := oleutil.GetProperty(it, name)
	if err != nil {
		return ""
	}
	return strings.TrimSpace(fmt.Sprintf("%v", v.Value()))
}

func intProp(it *ole.IDispatch, name string) int64 {
	v, err := oleutil.GetProperty(it, name)
	if err != nil {
		return 0
	}
	s := fmt.Sprintf("%v", v.Value())
	n, _ := strconv.ParseInt(strings.TrimSpace(s), 10, 64)
	return n
}

// ---- CPU ----

type CPUInfo struct {
	Name    string `json:"name"`
	Cores   int64  `json:"cores"`
	Threads int64  `json:"threads"`
	MaxMHz  int64  `json:"max_speed"`
	L3KB    int64  `json:"l3_cache"`
}

func GetCPUInfo() CPUInfo {
	wmi, service, err := wmiConnect()
	if err != nil {
		return CPUInfo{}
	}
	defer wmi.Release()
	defer service.Release()
	defer ole.CoUninitialize()

	items, cleanup := wmiQueryItems(service, "SELECT Name,NumberOfCores,NumberOfLogicalProcessors,MaxClockSpeed,L3CacheSize FROM Win32_Processor")
	defer cleanup()
	if len(items) == 0 {
		return CPUInfo{}
	}
	it := items[0]
	return CPUInfo{
		Name:    strProp(it, "Name"),
		Cores:   intProp(it, "NumberOfCores"),
		Threads: intProp(it, "NumberOfLogicalProcessors"),
		MaxMHz:  intProp(it, "MaxClockSpeed"),
		L3KB:    intProp(it, "L3CacheSize"),
	}
}

// ---- RAM ----

type RAMSlot struct {
	CapacityGB   float64 `json:"capacity_gb"`
	Speed        int64   `json:"speed"`
	Manufacturer string  `json:"manufacturer"`
	PartNumber   string  `json:"part_number"`
	Serial       string  `json:"serial"`
}

func GetRAMInfo() []RAMSlot {
	wmi, service, err := wmiConnect()
	if err != nil {
		return nil
	}
	defer wmi.Release()
	defer service.Release()
	defer ole.CoUninitialize()

	items, cleanup := wmiQueryItems(service,
		"SELECT Capacity,Speed,Manufacturer,PartNumber,SerialNumber FROM Win32_PhysicalMemory")
	defer cleanup()

	slots := make([]RAMSlot, 0, len(items))
	for _, it := range items {
		capB, _ := strconv.ParseInt(strProp(it, "Capacity"), 10, 64)
		slots = append(slots, RAMSlot{
			CapacityGB:   math.Round(float64(capB)/(1024*1024*1024)*100) / 100,
			Speed:        intProp(it, "Speed"),
			Manufacturer: strProp(it, "Manufacturer"),
			PartNumber:   strProp(it, "PartNumber"),
			Serial:       strProp(it, "SerialNumber"),
		})
	}
	return slots
}

// ---- Storage ----

type DiskInfo struct {
	Model       string  `json:"model"`
	SizeGB      float64 `json:"size_gb"`
	Serial      string  `json:"serial"`
	Interface   string  `json:"interface"`
	MediaType   string  `json:"media_type"`
}

func GetStorageInfo() []DiskInfo {
	wmi, service, err := wmiConnect()
	if err != nil {
		return nil
	}
	defer wmi.Release()
	defer service.Release()
	defer ole.CoUninitialize()

	items, cleanup := wmiQueryItems(service,
		"SELECT Model,Size,SerialNumber,InterfaceType,MediaType FROM Win32_DiskDrive")
	defer cleanup()

	disks := make([]DiskInfo, 0, len(items))
	for _, it := range items {
		sizeB, _ := strconv.ParseInt(strProp(it, "Size"), 10, 64)
		disks = append(disks, DiskInfo{
			Model:     strProp(it, "Model"),
			SizeGB:    math.Round(float64(sizeB)/(1024*1024*1024)*100) / 100,
			Serial:    strProp(it, "SerialNumber"),
			Interface: strProp(it, "InterfaceType"),
			MediaType: strProp(it, "MediaType"),
		})
	}
	return disks
}

// ---- GPU ----

type GPUInfo struct {
	Name          string  `json:"name"`
	DriverVersion string  `json:"driver_version"`
	RAMMB         float64 `json:"ram_mb"`
}

func GetGPUInfo() []GPUInfo {
	wmi, service, err := wmiConnect()
	if err != nil {
		return nil
	}
	defer wmi.Release()
	defer service.Release()
	defer ole.CoUninitialize()

	items, cleanup := wmiQueryItems(service,
		"SELECT Name,DriverVersion,AdapterRAM FROM Win32_VideoController")
	defer cleanup()

	gpus := make([]GPUInfo, 0, len(items))
	for _, it := range items {
		ramB, _ := strconv.ParseInt(strProp(it, "AdapterRAM"), 10, 64)
		gpus = append(gpus, GPUInfo{
			Name:          strProp(it, "Name"),
			DriverVersion: strProp(it, "DriverVersion"),
			RAMMB:         math.Round(float64(ramB) / (1024 * 1024)),
		})
	}
	return gpus
}

// ---- Motherboard ----

type MotherboardInfo struct {
	Manufacturer string `json:"manufacturer"`
	Product      string `json:"product"`
	Version      string `json:"version"`
}

func GetMotherboardInfo() MotherboardInfo {
	wmi, service, err := wmiConnect()
	if err != nil {
		return MotherboardInfo{}
	}
	defer wmi.Release()
	defer service.Release()
	defer ole.CoUninitialize()

	items, cleanup := wmiQueryItems(service,
		"SELECT Manufacturer,Product,Version FROM Win32_BaseBoard")
	defer cleanup()
	if len(items) == 0 {
		return MotherboardInfo{}
	}
	it := items[0]
	return MotherboardInfo{
		Manufacturer: strProp(it, "Manufacturer"),
		Product:      strProp(it, "Product"),
		Version:      strProp(it, "Version"),
	}
}

// ---- BIOS ----

type BIOSInfo struct {
	Version string `json:"version"`
	Serial  string `json:"serial"`
}

func GetBIOSInfo() BIOSInfo {
	wmi, service, err := wmiConnect()
	if err != nil {
		return BIOSInfo{}
	}
	defer wmi.Release()
	defer service.Release()
	defer ole.CoUninitialize()

	items, cleanup := wmiQueryItems(service,
		"SELECT SMBIOSBIOSVersion,SerialNumber FROM Win32_BIOS")
	defer cleanup()
	if len(items) == 0 {
		return BIOSInfo{}
	}
	it := items[0]
	return BIOSInfo{
		Version: strProp(it, "SMBIOSBIOSVersion"),
		Serial:  strProp(it, "SerialNumber"),
	}
}

// ---- Monitors ----

type MonitorInfo struct {
	Serial       string `json:"serial"`
	Model        string `json:"model"`
	Manufacturer string `json:"manufacturer"`
	SizeCM       int    `json:"size_cm"`
	Resolution   string `json:"resolution"`
}

func GetMonitors() []MonitorInfo {
	// We use WmiMonitorID from root\wmi namespace for model/serial
	if err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED); err != nil {
		// ok
	}
	defer ole.CoUninitialize()

	unknown, err := oleutil.CreateObject("WbemScripting.SWbemLocator")
	if err != nil {
		return nil
	}
	defer unknown.Release()
	wmiDisp, _ := unknown.QueryInterface(ole.IID_IDispatch)
	defer wmiDisp.Release()

	// Connect to root\wmi namespace
	serviceRaw, err := oleutil.CallMethod(wmiDisp, "ConnectServer", nil, "root\\wmi")
	if err != nil {
		return nil
	}
	service := serviceRaw.ToIDispatch()
	defer service.Release()

	monitors := []MonitorInfo{}

	// Get monitor IDs
	items, cleanup := wmiQueryItems(service, "SELECT ManufacturerName,UserFriendlyName,SerialNumberID FROM WmiMonitorID")
	defer cleanup()

	// Get display size params
	paramItems, paramCleanup := wmiQueryItems(service,
		"SELECT MaxHorizontalImageSize,MaxVerticalImageSize FROM WmiMonitorBasicDisplayParams")
	defer paramCleanup()

	// Get resolution from standard WMI
	var resolution string
	{
		wmi2, service2, err2 := wmiConnect()
		if err2 == nil {
			defer wmi2.Release()
			defer service2.Release()
			resItems, resCleanup := wmiQueryItems(service2,
				"SELECT CurrentHorizontalResolution,CurrentVerticalResolution FROM Win32_VideoController")
			defer resCleanup()
			if len(resItems) > 0 {
				h := intProp(resItems[0], "CurrentHorizontalResolution")
				v := intProp(resItems[0], "CurrentVerticalResolution")
				if h > 0 && v > 0 {
					resolution = fmt.Sprintf("%dx%d", h, v)
				}
			}
		}
	}

	for i, it := range items {
		// Decode byte-array properties (WMI returns as []uint16 / []interface{})
		model := decodeWMIByteArray(it, "UserFriendlyName")
		if model == "" {
			continue
		}
		serial := decodeWMIByteArray(it, "SerialNumberID")
		manufacturer := decodeWMIByteArray(it, "ManufacturerName")
		if serial == "" {
			serial = "N/A"
		}

		sizeCM := 0
		if i < len(paramItems) {
			w := int(intProp(paramItems[i], "MaxHorizontalImageSize"))
			h := int(intProp(paramItems[i], "MaxVerticalImageSize"))
			if w > 0 && h > 0 {
				sizeCM = int(math.Round(math.Sqrt(float64(w*w + h*h))))
			}
		}

		monitors = append(monitors, MonitorInfo{
			Serial:       serial,
			Model:        model,
			Manufacturer: manufacturer,
			SizeCM:       sizeCM,
			Resolution:   resolution,
		})
	}
	return monitors
}

// decodeWMIByteArray reads a WMI byte-array property (e.g. UserFriendlyName) as a string.
// WMI returns these as a SAFEARRAY of uint16 values (the raw name chars), terminated by 0.
func decodeWMIByteArray(it *ole.IDispatch, propName string) string {
	v, err := oleutil.GetProperty(it, propName)
	if err != nil {
		return ""
	}
	// The value comes back as []interface{} where each element is uint16 or int
	raw, ok := v.Value().([]interface{})
	if !ok {
		return ""
	}
	b := make([]byte, 0, len(raw))
	for _, ch := range raw {
		var n uint16
		switch val := ch.(type) {
		case uint16:
			n = val
		case int16:
			n = uint16(val)
		default:
			continue
		}
		if n == 0 {
			break
		}
		b = append(b, byte(n))
	}
	return strings.TrimSpace(string(b))
}
